src/test/java/com/driving/school/service/CrudServiceTest.java

```diff
@@ -33,34 +42,283 @@ class CrudServiceTest {
+        //and
+        setUpMock_repo_findAll(dbContent);
+
+        //when
+        List<ResponseDto> actual = service.getAll();
+
+        //then
+        verify(repo).findAll();
+        assertAllDtosAreCorrect(responseMappings, actual);
+    }
+
+    @Test
+    @DisplayName("getAll(): on an empty DB, returns empty list")
+    void whenGetAll_thenEmptyList() {
+        //given
+        setUpMock_repo_findAll(new ArrayList<>());
+
+        //when
+        List<ResponseDto> actual = service.getAll();
+
+        //then
+        verify(repo).findAll();
+        assertNoDtoWasReturned(actual);
+    }
+
+    @Test
+    @DisplayName("getById(): gets an entity")
+    void whenGetById_thenOK() {
+        //given
+        String foundEntity = setUpMock_repo_findById();
+        //and
+        StubResponseDto expectedResponseMapping = setUpMock_mapper_toResponseDto(foundEntity);
+
+        //when
+        ResponseDto actual = service.getById(1L);
+
+        //then
+        verify(repo).findById(anyLong());
+        assertCorrectDtoIsReturned(expectedResponseMapping, actual);
+    }
+
+    @Test
+    @DisplayName("patchById(): updates an existing entity")
+    void whenPatchById_thenOK() {
+        //given
+        setUpMock_repo_findById();
+        //and
+        String patchedEntity = setUpMock_repo_save();
+        //and
+        StubResponseDto expectedResponseMapping = setUpMock_mapper_toResponseDto(patchedEntity);
+        //and
+        //anonymous empty implementation (notice the {})
+        UpdateDto requestData = new UpdateDto() {};
+
+        //when
+        ResponseDto actual = service.patchById(1L, requestData);
+
+        //then
+        verify(repo).findById(anyLong());
+        assertCorrectDtoIsReturned(expectedResponseMapping, actual);
+    }
+
+    @Test
+    @DisplayName("create(): creates and saves an entity")
+    void whenCreate_thenOK() {
+        //given
+        setUpMock_mapper_toModel();
+        //and
+        String savedEntity = setUpMock_repo_save();
+        //and
+        StubResponseDto expectedResponseMapping = setUpMock_mapper_toResponseDto(savedEntity);
+        //and
+        //anonymous empty implementation (notice the {})
+        CreationDto requestData = new CreationDto() {};
+
+        //when
+        ResponseDto actual = service.create(requestData);
+
+        //then
+        verify(repo).save(any());
+        assertCorrectDtoIsReturned(expectedResponseMapping, actual);
+    }
+
+    @Test
+    @DisplayName("deleteById(): deletes existing entity")
+    void whenDeleteById_thenOK() {
+        //given
+        String savedEntity = setUpMock_repo_findById();
+
+        //when
+        service.deleteById(1L);
+
+        //then
+        verify(repo).findById(anyLong());
+        verify(removalUtil).deleteEntity(savedEntity);
+    }
+
+    @Test
+    @DisplayName("getById(): throws if ID doesn't exist")
+    void whenGetById_thenThrows() {
```

Comment:
- Te trzy metody `when*ByID_ThenThrows()` -  czy jest nam potrzebne testowanie każdej z tych metod jeśli tak naprawdę wykorzystujemy metodę `findById` i tu by wystarczyło tylko napisać test ```findById(): throws on invalid ID```, czy coś w tym stylu?

Reply:
W unit teście metody `findById()` na pewno tak zrobię, ale tu testuję serwis.

Reply:
Jeszcze uściślę: faktycznie pojawia się takie podejście, żeby unit testów robić trochę mniej. Kiedyś się testów nie robiło w ogóle, potem się robiło do wszystkiego, a teraz z unitów się niektórych rezygnuje, za to IT robi się równie pieczołowicie co właściwy kod.

Więc jeśli chodziło Ci o to, żeby zrezygnować z testowania tego zachowania w warstwie serwisu i zrobić tylko test w warstwie repo, to OK, można tak podejść. Problem tutaj jest tylko taki, że _rzucenie wyjątku_ jest odpowiedzialnością serwisu (nie repo) i dlatego właśnie w tej warstwie powinniśmy to testować.